### ISSUE TITLE
Mnemonics in menus

### ISSUE THREAD

**Issue by beastie1** [OtterBrowser/otter-browser#510]
I haven't found any mention of mnemonics (the '&' shortcuts in menus) anywhere. Would you consider adding them manually (e.g. **Enable Java&Script** in the **F12** menu) or do you prefer implementing some kind of automated system like in Opera, e.g.

```
&Open all pop-ups
O&pen pop-ups in background
&Block unwanted pop-ups
B&lock all pop-ups
```

Personally I'd prefer the first option as it's 1. straightforward to implement (just add an ampersand and you're good to go) and 2. more user-friendly (**Enable &Plugins** is much easier to remember than **E&nable Plug-Ins**).

**Comment by komoku** [comment]
I think @ersi-dnd makes some fair points. But to address points 1 and 2, why not write a script that generates the mnemonics automatically on the language files, and overwrite them, so that translators in Transifex have a baseline?

I.e., if the English language file has "New Tab", this script would change it into "&New Tab". Then the new language file would be updated to Transifex, and give two huge benefits to translators:

- It's obvious which strings correspond to menu items and which don't.
- There is an automatic baseline that can be easily editted only when needed, i.e., if a mnemonic is OK there's no need to change it, if a couple of mnemonics are not OK then changing those is enough.

By the way, I think there should be a more intelligent way of artificially generating mnemonics than the one Opera uses. Opera just seems to have a very simple rule such that it takes the 1st letter in the string if it's not already taken by other menu item, if it is, it goes to the 2nd, 3rd, etc. I think it's not that hard to come up with a better solution, e.g., if instead of just trying the characters in the string in order you first try first characters of words that are in uppercase, then first characters of words in lowercase, and then the rest, you'd get rid of awkward cases like the one mentioned by @beastie1. I suspect some decent enough set of heuristics (stuff like E&xit can come from taking into account letter frequency) would render manual mnemonics unnecessary in 99.9% of cases.

**Comment by pmattern** [comment]
#616 which was marked as blocker of this one got closed more than half a year ago. So is there any chance to get ahead now?
I'm aware of the problems stated by @ersi-dnd (and have been involved in translations myself already). But those shouldn't prevent a feature as important as this from getting implemented, in particular as Otter Browser isn't the only project which has to cope with them.

**Comment by ersi-dnd** [comment]
I agree that the feature is important. I also agree that other apps have it. I don't agree that it is easy to make. I think it is difficult to make it right. 

For example look at this http://i.imgur.com/w8cJ2gD.png It has "s" highlighted twice and therefore the access keys are unusable. If this thing were freely configurable for the user, then users could work around mistakes like this, but in this case it's not configurable and I have to use another language for this app.

Opera did it right. Opera had language files with all the strings. In this file, *some* strings had a &-sign inserted which made the menus have mnemonics where applicable. Users could freely reassign the mnemonics, create more of them or get rid of them. 

If this kind of solution is not available for Otter, I don't see the access keys as a good idea. 

Another idea that I would accept is an automatic access key generator for menus at build stage - with an option to not use it.

**Comment by Frenzie** [comment]
> For example look at this http://i.imgur.com/w8cJ2gD.png It has "s" highlighted twice and therefore the access keys are unusable.

Looks to me like lowercase `s` and uppercase `S`? Not a difference commonly made in these matters, admittedly, but one that makes sense to me.

To me the bigger question is to what extent automatically generated access keys *based on the actual text of the menu entry* make sense at all, and whether it might not be preferable to follow a simple linear algorithm such as [this](https://github.com/Frenzie/nimbler/blob/0eae4889751c8dea9c6e0157b929b64e498bad64/nimbler.py#L46). (Incidentally, there I do differentiate between uppercase and lowercase.) Unfortunately that would be non-standard and visually one can easily imagine it being too cluttered, or at least confusing to those who didn't know what it meant, unless the global shortcuts didn't show up in the menu and were replaced by these linearly generated ones. The big downside of traditional access keys like in Opera's bookmarks menu is that you often have to sort of hunt for them randomly in the middle of a word. But I think that's still much to be preferred over Firefox' method where sometimes you have an access key, sometimes you don't.

**Comment by ersi-dnd** [comment]
> Looks to me like lowercase  s  and uppercase  S ? Not a difference commonly made in these matters, admittedly, but one that makes sense to me.

Yes, it looks like that, but it's not. In that dialogue, you can use single-press keys or you can use them with Shift, and they both give the same result. Shift-s and single s both only access the first button where s is highlighted.

**Comment by Chocimier** [comment]
I started to work on this, but got some problems, e. g. with actions that changes text like Reload/Stop, so automatically generated shortcuts changes. Very-early version to play with is here: https://github.com/Chocimier/otter/tree/iss510 (not even shortcuts in menu bar, only in its menus).

**Comment by Zero3K** [comment]
How's this issue coming along?

**Comment by AluminumGriffin** [comment]
In 0.9.11-dev at least one can add hotkeys/accelerators by messing with the translation-files (.ts), just add &amp;amp; (ampersand-a-m-p-semicolon) before whatever key you want as an accelerator. (works when converting to a .qm with lconvert at least). (Takes effect when you switch language to the new .qm created)
